src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -842,21 +842,42 @@ impl<'tcx> IntRange<'tcx> {
         }
     }
 
+    #[inline]
+    fn integral_size_and_signed_bias(tcx: TyCtxt<'tcx>, ty: Ty<'_>) -> Option<(Size, u128)> {
+        match ty.kind {
+            ty::Char => Some((Size::from_bytes(4), 0)),
+            ty::Int(ity) => {
+                let size = Integer::from_attr(&tcx, SignedInt(ity)).size();
+                Some((size, 1u128 << (size.bits() as u128 - 1)))
+            }
+            ty::Uint(uty) => Some((Integer::from_attr(&tcx, UnsignedInt(uty)).size(), 0)),
+            _ => None,
+        }
+    }
+
     #[inline]
     fn from_const(
         tcx: TyCtxt<'tcx>,
         param_env: ty::ParamEnv<'tcx>,
         value: &Const<'tcx>,
     ) -> Option<IntRange<'tcx>> {
-        if Self::is_integral(value.ty) {
+        if let Some((target_size, bias)) = Self::integral_size_and_signed_bias(tcx, value.ty) {
```

Comment:
I was gonna say this can use the `?` operator, but not sure if that would kill perf again

Reply:
Not sure. It would also make `from_const()` look different to `from_range()`, which would be a shame.